src/device/intrinsics/random.jl

```diff
@@ -0,0 +1,49 @@
+function get_global_thread_index()
+    block_id = (gridDim().x * (blockIdx().y - 1)) + (blockIdx().x - 1)
+
+    thread_id = (block_id * (blockDim().x * blockDim().y)) +
+                ((threadIdx().y - 1) * blockDim().x) +
+                (threadIdx().x - 1)
+
+    return thread_id + 1
+end
+
+function xorshift(x::UInt32)::UInt32
+    x = xor(x, x << 13)
+    x = xor(x, x >> 17)
+    x = xor(x, x << 5)
+    return x
+end
+
+@eval @inline global_random_state() =
+    Base.llvmcall(
+        $("""@global_random_state = weak externally_initialized global i$(WORD_SIZE) 0
+             define i64 @entry() #0 {
+                 %ptr = load i$(WORD_SIZE), i$(WORD_SIZE)* @global_random_state, align 8
+                 ret i$(WORD_SIZE) %ptr
+             }
+             attributes #0 = { alwaysinline }
+          """, "entry"), Ptr{Cvoid}, Tuple{})
+
+function rand()
+    global_state_address = global_random_state()
+    global_state_ptr = LLVMPtr{UInt32, AS.Generic}(global_state_address)
+    index = get_global_thread_index()
+
+    state = unsafe_load(global_state_ptr, index, Val(8))
+    new_state = xorshift(state)
+    unsafe_store!(global_state_ptr, new_state, index, Val(8))
+
+    res = (new_state >> 9) | reinterpret(UInt32, 1f0)
+    return reinterpret(Float32, res) - 1f0
+end
+
+function init_random_state!(mod::CuModule, len::Int64, rng_seed)
+    random_state_ptr = CuGlobal{Ptr{Cvoid}}(mod, "global_random_state")
+    random_state = Mem.alloc(Mem.Device, sizeof(UInt32) * len)
+    rng = MersenneTwister(rng_seed)
+    seeds = Base.rand(rng, UInt32, len)
+    unsafe_copyto!(convert(CuPtr{UInt32}, random_state), pointer(seeds), len)
```

Comment:
Can we avoid the seed, e.g., by mixing in the thread id in the `rand()` function. Not sure how safe that is wrt. quality of randomness, but it would reduce launch overhead. If not, the copy needs to be async.